lib/runner.js

```diff
@@ -564,10 +564,16 @@ internals.protect = function (item, state, callback) {
     setImmediate(() => {
 
         domain.enter();
-        item.fn.call(null, (err) => {
+        const itemResult = item.fn.call(null, (err) => {
 
             finish(err, 'done');
         });
+        if (itemResult && itemResult.then instanceof Function) {
+            itemResult.then(() => finish(null), (err) => finish(err, 'done'));
```

Comment:
Thought it was obvious but you don't really need null either, you could just .then(finish ;)